File: template-method/typescript/TemplateMethod.ts
// Abstract class (template)
abstract class Beverage {
  prepareRecipe(): string[] {
    const results: string[] = [];
    results.push('Boiling water');
    results.push(this.brew());
    results.push('Pouring into cup');
    results.push(this.addCondiments());
    return results;
  }

  abstract brew(): string;
  abstract addCondiments(): string;
}

// Concrete implementations
class Tea extends Beverage {
  brew(): string {
    return 'Steeping the tea';
  }

  addCondiments(): string {
    return 'Adding lemon';
  }
}

class Coffee extends Beverage {
  brew(): string {
    return 'Dripping coffee through filter';
  }

  addCondiments(): string {
    return 'Adding sugar and milk';
  }
}

// Usage
console.log('Making tea:');
const tea = new Tea();
tea.prepareRecipe().forEach(step => console.log(`- ${step}`));

console.log('\nMaking coffee:');
const coffee = new Coffee();
coffee.prepareRecipe().forEach(step => console.log(`- ${step}`));

File: template-method/java/TemplateMethod.java
import java.util.ArrayList;
import java.util.List;

// Abstract class (template)
abstract class Beverage {
    public final List<String> prepareRecipe() {
        List<String> results = new ArrayList<>();
        results.add("Boiling water");
        results.add(brew());
        results.add("Pouring into cup");
        results.add(addCondiments());
        return results;
    }

    protected abstract String brew();
    protected abstract String addCondiments();
}

// Concrete implementations
class Tea extends Beverage {
    @Override
    protected String brew() {
        return "Steeping the tea";
    }

    @Override
    protected String addCondiments() {
        return "Adding lemon";
    }
}

class Coffee extends Beverage {
    @Override
    protected String brew() {
        return "Dripping coffee through filter";
    }

    @Override
    protected String addCondiments() {
        return "Adding sugar and milk";
    }
}

// Main class demonstrating Template Method pattern
public class TemplateMethod {
    public static void main(String[] args) {
        System.out.println("Making tea:");
        Beverage tea = new Tea();
        for (String step : tea.prepareRecipe()) {
            System.out.println("- " + step);
        }

        System.out.println("\nMaking coffee:");
        Beverage coffee = new Coffee();
        for (String step : coffee.prepareRecipe()) {
            System.out.println("- " + step);
        }
    }
}
